flow/core/kernel/vehicle/traci.py

```diff
@@ -130,8 +182,72 @@ def update(self, reset):
             specifies whether the simulator was reset in the last simulation
             step
         """
-        # copy over the previous speeds
+        # =================================================================== #
+        # Add the inflow vehicles.                                            #
+        # =================================================================== #
+        self._total_time += 1
+
+        for edge in self._inflows_by_edge.keys():
+            # This inflow is using a sumo-specific feature, so ignore.
+            if len(self._inflows_by_edge[edge]["cumsum"]) == 0:
+                continue
 
+            veh_per_hour = self._inflows_by_edge[edge]["cumsum"][-1]
+            steps_per_veh = 3600 / (self.sim_step * veh_per_hour)
+
+            # Add a vehicle if the inflow rate requires it.
+            if steps_per_veh < 1 or self._total_time % int(steps_per_veh) == 0:
+                # Choose the type of vehicle to push to this edge
+                names = self._inflows_by_edge[edge]["type"]
+                name = names[0]
+                cumsum = self._inflows_by_edge[edge]["cumsum"]
+                for i in range(len(names) - 1):
+                    # Deal with cases with no vehicles.
+                    if self._num_inflows[names[i]] == 0:
+                        break
+
+                    # This is used in order to maintain the inflow rate ratio.
+                    exp_inflow_ratio = (cumsum[i+1] - cumsum[i]) / cumsum[i]
+                    act_inflow_ratio = self._num_inflows[names[i+1]] \
+                        / sum(self._num_inflows[names[j]] for j in range(i+1))
+
+                    # If not enough vehicles of a specific type has been pushed
+                    # to the network yet, add it to the network.
+                    if exp_inflow_ratio < act_inflow_ratio:
+                        break
+                    else:
+                        name = names[i + 1]
+
+                # Choose the departure speed.
+                depart_speed = self._inflows[name]["departSpeed"]
+
+                # number of vehicles to add
+                num_vehicles = max(
+                    1,
+                    # number of vehicles to add if inflow rate is greater than
+                    # the simulation update time per step
+                    math.floor(1/steps_per_veh)
+                    + (1 if random.uniform(0, 1) <
+                       ((1/steps_per_veh) - math.floor(1/steps_per_veh))
+                       else 0)
+                )
+
+                for veh_num in range(num_vehicles):
```

Comment:
at each time step, we should shuffle the vehicles to be added so they're not added in batches (i.e. 5 humans, then 2 avs)